lib/PreviewProvider.js

```diff
@@ -0,0 +1,239 @@
+    */
+  cleanUpCache: Task.async(function*() {
+    let currentTime = Date.now();
+    Object.keys(ss.storage.embedlyData).forEach(item => {
+      if (currentTime - ss.storage.embedlyData[item].accessTime > CLEAR_CACHE_TIMEOUT) {
+        delete ss.storage.embedlyData[item];
+      }
+    });
+    this._cacheTimeout();
+  }),
+
+  /**
+    * Set up the timer for the cache to be cleaned every 24 hours
+    */
+  _cacheTimeout() {
+    this._timeoutID = setTimeout(() => {
+      this.cleanUpCache();
+    }, CLEAR_CACHE_TIMEOUT);
+  },
+
+  /**
+    * Filter out sites that do not have acceptable protocols and hosts
+    */
+  _URLFilter(definition) {
+    return function(item) {
+      let result = true;
+      definition.forEach((test) => {
+        if (!test(item)) {
+          result = false;
+        }
+      });
+      return result;
+    };
+  },
+
+  /**
+    * Santize the URL to remove any unwanted or sensitive information about the link
+    */
+  _sanitizeURL(site) {
+    if (!site) {
+      return "";
+    }
+
+    let queryParams = new Map();
+    let safeQueryParam = "";
+    let url = new URL(site);
+
+    if (url.search.length === 0) {
+      return url.toString();
+    }
+    // extract and parse the query parameters, if any
+    url.search.slice(1, url.search.length);
+    let paramArray = url.search.slice(1, url.search.length).split("&");
+    paramArray.forEach(item => {
+      let tmp = item.split("=");
+      queryParams.set(tmp[0], tmp[1]);
+    });
+    // remove the unwanted parameters and update the query string
+    REMOVE_KEYS.forEach(key => queryParams.delete(key));
+    queryParams.forEach((val, key) => {
+      if (val.length !== 0 && key.length !== 0) {
+        safeQueryParam = safeQueryParam + key + "=" + val + "&";
+      }
+    });
+    url.search = safeQueryParam.slice(0, safeQueryParam.length - 1);
+    url.hash = "";
+
+    // remove extra slashes in pathname and construct back a safe pathname
+    if (url.pathname) {
+      let pathItems = url.pathname.split("/");
+      let safePathItems = pathItems.filter(item => item.length !== 0);
+      let safePath = "/";
+      url.pathname = safePath + safePathItems.join("/");
+    }
+
+    // if the url doesn't have a protocol, use https
+    if (!url.protocol) {
+      url.protocol = "https:";
+    }
+
+    return url.toString();
+  },
+
+  /**
+    * Create a key based on a URL in order to dedupe sites
+    */
+  _createKey(url) {
+    url = new URL(url);
+    let key = url.host.replace(/www\.?/, "");
+    key = key + url.pathname;
+    return key;
+  },
+
+  /**
+    * Canonicalize urls by deduping them, then sanitizing them
+    */
+  _dedupeURLs(sites) {
+    let dedupedSites = new Map();
+    sites.forEach(site => {
+      let urlkey = this._createKey(site.url);
+      if (!dedupedSites.has(urlkey)) {
+        dedupedSites.set(urlkey, site);
+      } else {
+        // if two sites come in with 'http' and 'https', pick 'https'
```

Comment:
We saw that:

http://groupon.com redirects to https://www.groupon.com

Which means that the equivalent API response of http://embed.ly/docs/explore/extract?url=http%3A%2F%2Fgroupon.com returns an error.

I verified that there are actually two history items created in the example above, because Firefox considers redirects successful loads.

We can actually discriminate between the different kinds of loads:

 `sqlite> select p.url, v.visit_type, v.session from moz_historyvisits v, moz_places p where p.id = v.place_id order by v.visit_date DESC limit 5;`

```
url                                       visit_type  session
------------------------------------      ----------  ----------
https://www.groupon.com/                  5           0
http://www.groupon.com/                   2           0
```

We'll leave that as an exercise for later, however.

In most cases, a successful redirect is what we want.